### PR TITLE
[Clang] Introduce `CXXTypeidExpr::hasNullCheck`

### PR DESCRIPTION
**Pull request by MitalAshok**
Used to implement CWG2191 where `typeid` for a polymorphic glvalue only becomes potentially-throwing if the `typeid` operand was already potentially throwing or a `nullptr` check was inserted: https://cplusplus.github.io/CWG/issues/2191.html

Also change `Expr::hasSideEffects` for `CXXTypeidExpr` to check the operand for side-effects instead of always reporting that there are side-effects

Remove `IsDeref` parameter of `CGCXXABI::shouldTypeidBeNullChecked` because it should never return `true` if `!IsDeref` (we shouldn't add a null check that wasn't there in the first place)

### CODE REVIEW

**Comment by Endilll** [comment]
You need to run `clang/www/make_cxx_dr_status` script.

**Comment by Endilll** [comment]
If you update the branch, you should be able to get rid of unrelated changes to `cxx_dr_status.html`.

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -268,6 +268,9 @@ Resolutions to C++ Defect Reports
 - Clang now requires a template argument list after a template keyword.
   (`CWG96: Syntactic disambiguation using the template keyword <https://cplusplus.github.io/CWG/issues/96.html>`_).
 
+- Clang no longer always reports ``!noexcept(typeid(expr))`` when the ``typeid`` cannot throw a ``std::bad_typeid``.
```
Comment (Endilll) [inline clang/docs/ReleaseNotes.rst:271]:
```suggestion
- Clang now considers ``noexcept(typeid(expr))`` more carefully, instead of always assuming that ``std::bad_typeid`` can be thrown.
```

File: clang/www/cxx_dr_status.html [hunk 2]
```diff
@@ -15598,7 +15598,7 @@ <h2 id="cxxdr">C++ defect report implementation status</h2>
   </tr>
   <tr class="open" id="2632">
     <td><a href="https://cplusplus.github.io/CWG/issues/2632.html">2632</a></td>
-    <td>review</td>
+    <td>drafting</td>
```
Comment (Endilll) [inline clang/www/cxx_dr_status.html]: Update you branch with the latest `main` changes, and those unrelated changes will go away.

**Review by Endilll — APPROVED** [review]
DR test looks good, but wait for more approvals.

File: clang/lib/AST/ExprCXX.cpp [hunk 3]
```diff
@@ -166,6 +166,55 @@ QualType CXXTypeidExpr::getTypeOperand(ASTContext &Context) const {
       Operand.get<TypeSourceInfo *>()->getType().getNonReferenceType(), Quals);
 }
 
+namespace {
+static bool isGLValueFromPointerDeref(const Expr *E) {
```
Comment (Sirraide) [inline clang/lib/AST/ExprCXX.cpp:170]: Pretty sure `namespace {` *and* `static` is redundant; we usually only use `static`.

**Review by Sirraide — APPROVED** [review]
LGTM as well.
